File: src/Domain/Teams/State.fs
namespace Orleans.Tournament.Domain.Teams
open System

type TeamState =
    val mutable Id:Guid
    val mutable Created:bool
    val mutable Name:string
    val mutable Players:list<string>
    val mutable Tournaments:list<Guid>
    new() = {
        Id = Guid.Empty;
        Created = false;
        Name = String.Empty;
        Players = []
        Tournaments = []
    }
    new (id, created, name, players, tournaments) = {
        Id = id
        Created = created
        Name = name
        Players = players
        Tournaments = tournaments
    }
    member x.Apply (evt : TeamCreated) =
        x.Created <- true;
        x.Id <- evt.TeamId;
        x.Name <- evt.Name;
    member x.Apply (evt : PlayerAdded) =
        x.Players <- evt.Name :: x.Players
    member x.Apply (evt : TournamentJoined) =
        x.Tournaments <- evt.TournamentId :: x.Tournaments;

File: src/DomainC/Teams/State.cs
namespace Tournament.Domain.Teams;

public partial class TeamState
{
    public Guid Id { get; set; }
    public bool Created { get; set; }
    public string Name { get; set; }
    public List<string> Players { get; set; }
    public List<Guid> Tournaments { get; set; }

    public TeamState()
    {
        Name = string.Empty;
        Players = Enumerable.Empty<string>().ToList();
        Tournaments = Enumerable.Empty<Guid>().ToList();
    }

    public TeamState(
        Guid id,
        bool created,
        string name,
        List<string> players,
        List<Guid> tournaments)
    {
        Id = id;
        Created = created;
        Name = name;
        Players = players;
        Tournaments = tournaments;
    }

    public void Apply(TeamCreated evt)
    {
        Id = evt.TeamId;
        Name = evt.Name;
        Created = true;
    }

    public void Apply(PlayerAdded evt)
        => Players.Add(evt.Name);

    public void Apply(TournamentJoined evt)
        => Tournaments.Add(evt.TournamentId);
}